### PR TITLE
[WIP, FS-1043] Extension members visible to trait constraints

### PR DESCRIPTION
**Pull request by TobyShaw**
#### Language feature: Make extension members visible to trait constraints

Implements related language suggestions 
* https://github.com/fsharp/fslang-suggestions/issues/230
* https://github.com/fsharp/fslang-suggestions/issues/29

TODO items from the code

* [ ] TastOps.fs ``extSlns // TODO: do we need to remap here???``
* [ ] TastPickle.fs ``extSlns starts empty.  TODO: check the ramifications of this when inlining solved trait calls from other assemblies``
* [ ] Optimizer.fs ``// TODO: consider what happens when the expression refers to extSlns that have become hidden``
* [ ] Check use of ``None`` for TraitFreshener, e.g. https://github.com/Microsoft/visualfsharp/pull/3582/files#diff-5b9ab9dd9d7133aaf23add1048742031R576
* [ ] ConstraintSolver.fs ``// TODO: check the use of 'allPairs' - not all these extensions apply to each type variable.``

Things to test
* [ ] explicit instantiations of all the different possible generic things
* [ ] Using C#-style extension members defined in F# to satisfy constraints
* [ ] More uses of extension constraints

#### Old comments by Toby 

Obviously not merge-ready. Opening it up for discussion and feedback.

Seems to work in the simple examples I've tested.

I need to test it with more complex examples that originally motivated this patch.
I'm not happy with my strategy for passing around the information I need, I feel like it could be done in a more clean manner.

### CODE REVIEW

**Comment by abelbraaksma** [comment]
@TobyShaw, I'm curious, is this an improvement that would allow something like the following, for instance?

    module Test =
        type System.Int64 with
            static member MaxValue = 12

        let inline getMax< ^a when ^a: (static member MaxValue: ^a)> (x: ^a) = 
                let max = (^a: (static member MaxValue: ^a) ())
                max

        let testResult = getMax 12L

Currently that (righly, according to the docs) throws a compile-time exception:

image

BTW: this actually comes from a real-world scenario (see [this stackoverflow question on MinValue and MaxValue fields](https://stackoverflow.com/questions/40998545/using-numeric-generics-to-get-min-max-of-native-numeric-types-int32-maxvalue-do), since `MaxValue` only exists as field and static member constraints have another limitation: they don't apply to fields or constants. The only current workaround is to create wrapper types, with added complexities and runtime + GC overhead.

So, if your fix addresses that issue, it would be HUGE improvement! Even though it would still mean, of course, adding those fields as members, but [code like this](https://github.com/gusty/FsControl/blob/615f86cbce90596740efdc717cb863a57cefeee4/FsControl.Core/Numeric.fs#L476) (by @gusty) could well become way more readable and concise with such additions.

**Comment by TobyShaw** [comment]
I believe so yes, but I'll hold off on making any bold claims until I've done more testing.

**Comment by dsyme** [comment]
@TobyShaw If you could update to master that would be great (all the conflicts will be about spaces I think)

**Comment by dsyme** [comment]
@TobyShaw Great work :) I've added links to two language design suggestions that you should review, and add extra test cases for.

We will need to add some kind of access checking to allow internal members to be used in solving constraints, possibly in this PR - I'm not certain though

**Comment by TobyShaw** [comment]
Seems like you linked the same suggestion twice? I assume you meant [https://github.com/fsharp/fslang-suggestions/issues/243](https://github.com/fsharp/fslang-suggestions/issues/243)

**Comment by dsyme** [comment]
@TobyShaw I meant https://github.com/fsharp/fslang-suggestions/issues/29, thanks

File: src/fsharp/Fsc/Fsc.fsproj [hunk 1]
```diff
@@ -23,16 +22,28 @@
     <AllowCrossTargeting>true</AllowCrossTargeting>
     <OtherFlags>$(OtherFlags) --warnon:1182</OtherFlags>
   </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
```
Comment (saul) [inline src/fsharp/Fsc/Fsc.fsproj]: I believe this project file needs reverting, see the project.json changes below. I trust @dsyme can help you integrate some compiler unit tests

Comment (TobyShaw) [reply src/fsharp/Fsc/Fsc.fsproj]: Oops, didn't mean to commit this. Good spot

**Comment by gusty** [comment]
@TobyShaw Awesome, this is a very interesting feature and I consider it another step down the road to trait or typeclasses.

@abelbraaksma I agree with you. A project like [FSharpPlus](https://github.com/gusty/FSharpPlus) will benefit a lot with this feature, it will simplify a lot of the code used there internally and the resulting signatures as well.

@dsyme An interesting test will be to compile the library as it is now with this version of the compiler to check if there is a breaking change.
Then I will try to port the code to take advantage of this feature and see how far can be used, I mean if there is any limitation.

**Comment by TobyShaw** [comment]
Okay, done some testing. 

Many functions work perfectly as expected:
Append : ^A * ^A -> ^A
Show : ^A -> String
Incr : ^A -> ^A

Some functions don't: I don't get a unique overload, indicating a bug in my implementation, I reckon this can be fixed pretty easily:
Max : unit -> ^A

The two large problems are:

Indirections don't seem to be working currently, pretty huge problem which I need to look into.
let inline (++) a b = Append a b // Won't detect it any more

Doesn't currently work with extension members on generic types, causes the compiler to crash, I think this should be an easy fix though.

**Comment by dsyme** [comment]
@TobyShaw Could you merge master into your branch please? 

``git pull http://github.com/Microsoft/visualfsharp master`` + and resolve conflicts

**Comment by dsyme** [comment]
> The three large problems which make this useless at the moment are...
> 
> Indirections don't seem to be working currently, pretty huge problem which I need to look into.
> ```let inline (++) a b = Append a b // Won't detect it any more```

I'm not sure what you mean here - could you give a full example please?

> Doesn't work across Module boundaries. Although it can find the correct extension member when constraint solving, the NameResolutionEnv I pass on to the optimiser/code generator is only the one at the top level. Ie. I need to find a more localised means of reobtaining the correct extension member in CodegenWitnessThatTypSupportsTraitConstraint. Seems rather obvious that passing around the global NRE is a dumb idea, and this shows why.

Yes, we have to find a way to address this. 

> As a side: if someone could explain to me why we seem to do the constraint solving twice, that would be helpful. 

Basically because we don't correctly pass witness solutions all the way through.


> Doesn't currently work with extension members on generic types, causes the compiler to crash, I think this should be an easy fix though.

Yes, that should be an easy fix hopefully

**Comment by gusty** [comment]
@TobyShaw I worked before with the Contraint Solver and I also noticed that it solves twice (actually before my PRs it was solving it 4 times or more dependeing on the scenario).

But if my memory is not bad I stopped optimizing at the point I noticed that the CodegenWitnessThatTypSupportsTraitConstraint gives different results in some cases, that's why I didn't went further.

**Comment by TobyShaw** [comment]
@dsyme For an example of the indirection causing failures:

```
type System.Int32 with
  static member Append(a : System.Int32, b : System.Int32) = a + b

type System.Boolean with
  static member Append(a : System.Boolean, b : System.Boolean) = a && b

let inline append< ^A when ^A : (static member Append : ^A * ^A -> ^A) > (a : ^A) (b : ^A) =
  ( ^A : (static member Append : ^A * ^A -> ^A) (a,b) )

let inline append2 x y = append x y

[<EntryPoint>]
let main args =
  printfn "%A" (append 1 2)
  printfn "%A" (append2 true false)
  1
```

The first line (using append) works fine. The second line (using append2) fails, despite their equivalency.

**Comment by TobyShaw** [comment]
The merge conflict was pretty massive, may take a while to resolve

**Comment by dsyme** [comment]
> The first line (using append) works fine. The second line (using append2) fails, despite their equivalency.

OK, thanks, that does need to be fixed :)

**Comment by dsyme** [comment]
@TobyShaw I have merged with master and pushed to your branch.  Still building and ironing out issues with that

**Comment by dsyme** [comment]
@TobyShaw OK, this is merged now, and should now build correctly. I'm going to play with it a bit and develop a list of acceptance criteria

**Comment by TobyShaw** [comment]
Thanks for that! Sorry for the inactivity, it'll be at least a few days before I'm truly less busy and can work on this again.

**Comment by dsyme** [comment]
@TobyShaw I'm trying to find time to work on this - I've done a fair bit and will push to your repo/branch once it is stable.

**Comment by dsyme** [comment]
@TobyShaw This now works, to some extent.  See the testfile for example

Lots of things to iron out and many, many corner cases to test. Most important thing I noticed is that .NET/C#-defined extension members can't satisfy constraints - which would be great to have.

### example 1

```fsharp
type System.Int32 with
    static member (++)(a: int, b: int) = a 

1 ++ 2
```

### example 2

```fsharp
type System.Int32 with
    static member Add(a: int, b: int) = a 

type MyType =
    | MyType of int

[<AutoOpen>]
module Extensions = 
    type MyType with
        static member Add(MyType x, MyType y) = MyType (x + y)

let inline addGeneric< ^A when ^A : (static member Add : ^A * ^A -> ^A) > (a,b) : ^A =
    (^A : (static member Add : ^A * ^A -> ^A) (a,b))

let inline (+++) a b = addGeneric(a,b)

let inline addGeneric2  (a,b) : ^A when ^A : (static member Add : ^A * ^A -> ^A) =
    (^A : (static member Add : ^A * ^A -> ^A) (a,b))

let inline (++++) a b = addGeneric2(a,b)


let f () =
    let v1 = addGeneric (MyType(1), MyType(2))
    let v2 = addGeneric (1,1)
    ()

let f2 () =
    let v1 = MyType(1) +++ MyType(2)
    let v2 = 1 +++ 1
    1

let f3 () =
    let v1 = addGeneric2 (MyType(1), MyType(2))
    let v2 = addGeneric2 (1,1)
    ()

let f4 () =
    let v1 = MyType(1) ++++ MyType(2)
    let v2 = 1 ++++ 1
    ()  
```

**Comment by TobyShaw** [comment]
This is really cool, thanks for taking the time to help on this :)

It'll take some time to digest the changes you've made, I can see some of it was some necessary refactoring, but you've also added a lot of new things.

**Comment by dsyme** [comment]
@TobyShaw Yes, there are some important changes.  The most important things are

1. we scrape the set of possible solutions at point-of-use (when constraints are logically copied, e.g. when using a generic function) rather than point-of-syntax, and

2. we only force member trait constraint solving when the support of the constraint is nil - e.g. not when "permitWeakResolution" is set.

The second is a technical subtlety and I need to consider its ramifications more, but it was needed for the sampes

Will be continuing to work on this today

**Comment by dsyme** [comment]
@TobyShaw I did a lot of work on this and it's getting closer to being in shape - though there is still a _lot_ of testing to do - this thing has many interactions with other features.  

To give you a feel for what is working, the contents of testfiles/test.fs are now below

```fsharp
module Test

type MyType =
    | MyType of int

/// Extending a .NET primitive type with new operator
module DotNetPrimtiveWithNewOperator = 
    type System.Int32 with
        static member (++)(a: int, b: int) = a 

    let result = 1 ++ 2

/// Extending a .NET primitive type with new _internal_ operator
module DotNetPrimtiveWithInternalOperator1 = 
    type System.Int32 with
        static member internal (++)(a: int, b: int) = a 

    let result = 1 ++ 2 // this is now allowed


/// Extending a .NET primitive type with new _private_ operator where that operator is accessible at point-of-use
module DotNetPrimtiveWithAccessibleOperator2 = 
    type System.Int32 with
        static member private (++)(a: int, b: int) = a 

    let result = 1 ++ 2 // this is now allowed. 



#if NEGATIVE_TESTS
module DotNetPrimtiveWithInaccessibleOperator = 
    [<AutoOpen>]
    module Extensions = 
        type System.Int32 with
            static member private (++)(a: int, b: int) = a 

    let result = 1 ++ 2 // This should fail to compile because the private member is not accessible from here
#endif


/// Locally extending an F# type with a wide range of standard operators
module FSharpTypeWithExtrinsicOperators = 

    [<AutoOpen>]
    module Extensions = 
        type MyType with
            static member (+)(MyType x, MyType y) = MyType (x + y)
            static member (*)(MyType x, MyType y) = MyType (x * y)
            static member (/)(MyType x, MyType y) = MyType (x / y)
            static member (-)(MyType x, MyType y) = MyType (x - y)
            static member (~-)(MyType x) = MyType (-x)
            static member (|||)(MyType x, MyType y) = MyType (x ||| y)
            static member (&&&)(MyType x, MyType y) = MyType (x &&& y)
            static member (^^^)(MyType x, MyType y) = MyType (x &&& y)
            static member Zero = MyType 0
            static member One = MyType 1
            member this.Sign = let (MyType x) = this in sign x
            static member Abs (MyType x) = MyType (abs x)
            static member Sqrt (MyType x) = MyType (int (sqrt (float x)))
            static member Sin (MyType x) = MyType (int (sin (float x)))
            static member Cos (MyType x) = MyType (int (cos (float x)))
            static member Tan (MyType x) = MyType (int (tan (float x)))
            static member DivideByInt (MyType x, n: int) = MyType (x / n)

    let v = MyType 3
    let result1 = v + v  
    let result2 = v * v
    let result3 = v - v
    let result4 = v / v
    let result5 = -v
    let result6 = v ||| v
    let result7 = v &&& v
    let result8 = v ^^^ v
    let result9 = LanguagePrimitives.GenericZero<MyType>
    let result10 = LanguagePrimitives.GenericOne<MyType>
    let result11 = sign v
    let result12 = abs v
    let result13 = sqrt v
    let result14 = sin v
    let result15 = cos v
    let result16 = tan v
    let result17 = LanguagePrimitives.DivideByInt v 4


/// Extending two types with the static member 'Add'
module TwoTypesWithExtensionOfSameName = 

    [<AutoOpen>]
    module Extensions = 
        type System.Int32 with
            static member Add(a: int, b: int) = a 

        type MyType with
            static member Add(MyType x, MyType y) = MyType (x + y)

    let inline addGeneric< ^A when ^A : (static member Add : ^A * ^A -> ^A) > (a,b) : ^A =
        (^A : (static member Add : ^A * ^A -> ^A) (a,b))

    let inline (+++) a b = addGeneric(a,b)

    let inline addGeneric2  (a,b) : ^A when ^A : (static member Add : ^A * ^A -> ^A) =
        (^A : (static member Add : ^A * ^A -> ^A) (a,b))

    let inline (++++) a b = addGeneric2(a,b)


    let f () =
        let v1 = addGeneric (MyType(1), MyType(2))
        let v2 = addGeneric (1,1)
        ()

    let f2 () =
        let v1 = MyType(1) +++ MyType(2)
        let v2 = 1 +++ 1
        1

    let f3 () =
        let v1 = addGeneric2 (MyType(1), MyType(2))
        let v2 = addGeneric2 (1,1)
        ()

    let f4 () =
        let v1 = MyType(1) ++++ MyType(2)
        let v2 = 1 ++++ 1
        ()  


/// Extending a generic type with a property
module ExtendingGenericTypeWithProperty  = 

    type List<'T> with
        member x.Count = x.Length

    let inline count (a : ^A  when ^A : (member Count : int)) =
        (^A : (member Count : int) (a))

    let v0 = [3].Count // sanity check 

    let v3 = count [3]

    let v5 = count (ResizeArray [| 3 |])

/// Extending a generic type with a property
/// Extending the .NET array type with a property
module ExtendingGenericTypeAndArrayWithProperty  = 

    type List<'T> with
        member x.Count = x.Length

    type ``[]``<'T> with
        member x.Count = x.Length

    let inline count (a : ^A  when ^A : (member Count : int)) =
        (^A : (member Count : int) (a))

    let v0 = [3].Count // sanity check 

    let v1 = [|3|].Count // sanity check 

    let v3 = count [3]

    let v4 = count [| 3 |]

    let v5 = count (dict [| 1,3 |])

    let v6 = count (ResizeArray [| 3 |]) // intrinsic from .NET




/// Solving using LINQ extensions
module LinqExtensionMethodsProvideSolutions_Count = 

    open System.Linq

    // Note this looks for a _method_ called `Count` taking a single argument
    // It is _not_ considered the same as a proprty called `Count`
    let inline countm (a : ^A  when ^A : (member Count : unit -> int)) =
        (^A : (member Count : unit -> int) (a))

    let seqv : seq<int> = Seq.singleton 1
    
    let v0 = seqv.Count // sanity check 

    let v1 = countm seqv

/// A random example
module ContainsKeyExample   = 
    let inline containsKey (k: ^Key) (a : ^A  when ^A : (member ContainsKey : ^Key -> bool)) =
        (^A : (member ContainsKey : ^Key -> bool) (a,k))

    let v5 = containsKey 1 (dict [| 1,3 |])

    // Note that without 'inline' this doesn't become generic
    let inline f x = containsKey x (dict [| (x,1) |])

```

**Comment by gusty** [comment]
@dsyme I have a few questions:

What is expected when there are 2 extensions defined?
I guess the expected behaviour is to call the one corresponding to the last module which was opened.

Also, what happens if the method already exists?
Right now there is a strange behaviour with extensions that caused breaking changes in my code base when extensions methods I had defined in my library were added as proper methods in newer versions of F# core.

They simply stopped working and now they are silently ignored in favor of the one in core.
Here's an example of that issue: https://stackoverflow.com/questions/45363232/overriding-static-properties-in-f-with-extension-methods/46251858#46251858 

I know this is not necessarily related to this feature, but I think it's a good time to solve it, and make sure the same behaviour is observed when called through Static Constraints.

**Comment by dsyme** [comment]
> What is expected when the same extension is defined twice?
> Also, what happens if the method already exists?

The rules are the same as if you were making a method overload call at the point where a trait-qualified generic function is used. This can be subtle since the rules for extension method priorities are somewhat subtle, but basically
* Intrinsic (non-extension) things are preferred
* The most recently opened extension methods are preferred all else being equal

So this is ok (no ambiguity, use the most recently opened one)
```
    [<AutoOpen>]
    module Extensions = 
        type System.Int32 with
            static member (++)(a: int, b: int) = a 

    [<AutoOpen>]
    module Extensions2 = 
        type System.Int32 with
            static member (++)(a: int, b: int) = a 

    let result = 1 ++ 2

```
But this gives an ambiguity error:
```
    [<AutoOpen>]
    module Extensions = 
        type System.Int32 with
            static member (++)(a: int, b: int) = a 

    [<AutoOpen>]
    module Extensions2 = 
        type System.Int32 with
            static member (++)(a: int, b: string) = a 

    let f x = 1 ++ x
```
as does this:
```
    [<AutoOpen>]
    module Extensions = 
        type System.Int32 with
            static member (++)(a: int, b: int) = a 
            static member (++)(a: int, b: string) = a 

    let f x = 1 ++ x
```


> Right now there is a strange behaviour with extensions that caused breaking changes in my code base when extensions methods I had defined in my library were added as proper methods in newer versions of F# core. They simply stopped working and now they are silently ignored in favor of the one in core.
Here's an example of that issue: https://stackoverflow.com/questions/45363232/overriding-static-properties-in-f-with-extension-methods/46251858#46251858


Yes, it's true that's a property of extension methods.


> I know this is not necessarily related to this feature, but I think it's a good time to solve it, and make sure the same behaviour is observed when called through Static Constraints.

We would have o make any change orthogonal - the behaviour you describe is correct according to the language  spec.  But perhaps we should give a warning if an extension method is ignored.

**Comment by gusty** [comment]
> We would have o make any change orthogonal - the behaviour you describe is correct according to the language spec. But perhaps we should give a warning if an extension method is ignored.

@dsyme if ask me I would prefer to implement the override instead of the warning. Why? Because it's less restrictive but mainly because the current behavior is prone to breaking changes, as I described each time you add a method to a type in F# core you might be breaking a library or application that had that method before as an extension. As I said it already happened to me and will continue happening unless we change that situation both in the compiler and the spec.

**Comment by dsyme** [comment]
> @dsyme if ask me I would prefer to implement the override instead of the warning. Why? Because it's less restrictive but mainly because the current behavior is prone to breaking changes, as I described each time you add a method to a type in F# core you might be breaking a library or application that had that method before as an extension. As I said it already happened to me and will continue happening unless we change that situation both in the compiler and the spec.

However it would equally be a breaking change to prefer extension methods over original methods

**Comment by gusty** [comment]
> However it would equally be a breaking change to prefer extension methods over original methods

True, but:

1 - It will be only one-off, whereas right now each new version of F# core which adds a single method to a core type is a breaking change. I'm sure each new version of the core had and will have at least a new method, if so we can agree that each new release is a breaking change.

2 - It's very unlikely that this one-off breaking change actually happens, I mean unless someone has  code with extensions that never worked, never got executed and for some reason left that that code in the source, or just forgot to remove it after noticing that it didn't work.

PS: @dsyme let me know if you want to move this discussion to another place, we can include it in an RFC or open a suggestion if you prefer.

**Comment by dsyme** [comment]
> 1 - It will be only one-off, whereas right now each new version of F# core which adds a single method to a core type is a breaking change. 

Yes, but the same is true for every version of .NET that adds methods for C#.  We should do whatever C# does in this case (e.g. warn)

**Comment by gusty** [comment]
> We should do whatever C# does

Should we really follow them on every decision? Even if what they do is not the best approach?

In anycase we're not doing the same overload resultion as C#, the rules are different. Basic rules I mean.

Moreover, AFAIK they changed the overload resolution rules sometime ago.

@dsyme  I think we can do it better ;)

**Comment by dsyme** [comment]
> Should we really follow them on every decision? Even if what they do is not the best approach?

This is in the "decided in F# 1.0" category I'm afraid.  There are reasons that C# went with this decision. But regardless, it's orthogonal to what's in question here.  I think we could only change it through a gradual process of warning.

> 2 - It's very unlikely that this one-off breaking change actually happens, I mean unless someone has code with extensions that never worked, never got executed and for some reason left that that code in the source, or just forgot to remove it after noticing that it didn't work.

Well, I've seen it - it happens.

**Comment by abelbraaksma** [comment]
> > 2 - It's very unlikely that this one-off breaking change actually happens, I mean unless someone has code with extensions that never worked, never got executed and for some reason left that that code in the source, or just forgot to remove it after noticing that it didn't work.
>>
> Well, I've seen it - it happens.

While I sympathize with wanting to change this behavior, even more so since I have been bitten by it on more than one occasion, I have to agree with @dsyme here that it is one of those few areas where fixing an error situation can introduce subtle bugs in existing code.

For instance, I had the following:

```f#
type Single with
    /// An override to Single.Parse with fixed InvariantCulture
    static member Parse value = Single.Parse(value, floatNumberStyles, CultureInfo.InvariantCulture)
```

I never noticed this wasn't hit, and it wasn't a problem until the program was run on a non-US/UK system (which broke, because it never called the extension method).

Then I changed it thus (note the comment):

```f#
type Single with
    /// An override to Single.Parse with fixed InvariantCulture
    /// NOTE: cannot be renamed Single.Parse as it will *not* shadow its original.
    static member ParseInvariant value = Single.Parse(value, floatNumberStyles, CultureInfo.InvariantCulture)
```

@gusty, your proposal would have suddenly "fixed" my code. I would typically consider that _a good thing_, but not if that's the result of building on a different version of the F# compiler. The thing is that it would change the semantics of my code, and that's a much bigger surprise than you may think.

Though I would love to have the ability to _override_ existing signatures. I often jump through hoops, or have to invent new names or types. For instance, in the `Parse` example, it would be a great benefit if all my projects can use overridden methods `Double.Parse`, `Single.Parse`, etc. And I am sure other examples exist where such pattern would be a good thing.

_If_ that's going to be on the table, I would like to suggest to do that through a keyword, like `new` or `override`, which have similar semantics in OO (I have a slight preference for "new"):

```f#
type Single with
    /// An override to Single.Parse, with fixed InvariantCulture
    static member new Parse value = Single.Parse(value, floatNumberStyles, CultureInfo.InvariantCulture)
```

**Comment by gusty** [comment]
>  Compiling on F# 4.1 or F# 4.2 should not change semantics.

But today it does. With each new additional method included in the types defined in core, semantics are being changed all the time.

@abelbraaksma I think your example was exactly what I had in mind, but my conclusions are slightly different, or the way I "read" the situation.

Your code was a time-bomb, I mean it has something that will eventually show up and do something wrong.

For time-bombs, containing code that was inadvertently left, it will affect, but that situation will not be worst than changing semantics with every new version of the compiler and with code that in principle nobody can really tell it's a time-bomb, because in theory is not.

Let's put it this way, by doing a static analysis of your code we can detect that it has a defect. 

But by doing the same analysis on a code that is correct, we can't arrive to that conclusion, however a new version of core comes and suddenly our extension method stops executing and now it either breaks immediately or becomes a time-bomb. But there was no way to tell that the code was wrong, because nobody knows the future, nobody can assure that a new version of core will have a method with that name.

So if we accept this situation, **right now all code that defines extensions over primitive types is potentially a time-bomb**, because there is a small chance that a new version of core will have that method and then it will break, or at least as you say, it will change the semantics of your code.

Having said that and now trying to switch hats, I think it will be an interesting first step to have a version of the compiler which emits that warning and ask you to remove that piece of code because a new version of the compiler might fix that problem and that code will start executing.

And if we think about it, even if we convince @dsyme to accept that change, it won't be ready for F# vNext, but the warning can possible make it.

**Comment by abelbraaksma** [comment]
> Your code was a time-bomb, I mean it has something that will eventually show up and do something wrong.

Yes, and I exaggerated my situation a bit for dramatic effect, apologies for that 😉.

> Let's put it this way, by doing a static analysis of your code we can detect that it has a defect.

There we disagree (though that wasn't part of my point of view). A compiler cannot possibly determine it is defective, simply because the code is legal and the method can be used, albeit only with an FQN (fully-qualified name, i.e. including full namespace).

> because there is a small chance that a new version of core will have that method and then it will break, or at least as you say, it will change the semantics of your code.

Yes, this is certainly true. But there is one big difference and we should be careful not to compare apples and pears. When a new BCL method is introduced (doesn't quite matter whether that's a .NET BCL or FSharp.Core method), there's always a chance that an equal method exists. But that's for **one method with one signature and name**. 

The change you seem to be proposing applies to **all and any extension methods**. Even though the chance is tiny in real-life code out there, it means that people need to be educated / informed and that they need to change their coding habits. It's a language change vs. a library change. 

Btw, I'm not vehemently opposing your suggestion, in fact, I am in favor of it, but incompatible changes are hard to push through, if at all. Make it compatible and it's easier to get Microsoft on board.

Perhaps we should take the rest of this discussion to #3692? I agree that a warning is a good first step.

**Comment by robkuz** [comment]
Will this PR also adress this [comment](https://github.com/fsharp/fslang-suggestions/issues/230#issuecomment-300100999) about overloaded function with generic paramters?  
Or is it better to open an issue for that?

**Comment by jindraivanek** [comment]
I tried this PR, but I got errors on test samples (same errors I got if I run it on current fsc). If I revert to commit https://github.com/Microsoft/visualfsharp/pull/3582/commits/f287f20413f62e4cf65f0d9423e68c17b0c938b1, it works.

**Comment by jindraivanek** [comment]
Last commit that works is https://github.com/Microsoft/visualfsharp/pull/3582/commits/000e34d78e393ab6a4301ed548959ba009fcd8a2.

**Comment by realvictorprm** [comment]
Playing againg Ping Pong.

Pinging @TobyShaw and @dsyme. Let's get this ready to merge!

**Comment by realvictorprm** [comment]
OK @TobyShaw, I'll try to takeover then 🤔

**Comment by gusty** [comment]
@realvictorprm I think before merging this PR we should fix all current oddities with SRTP

**Comment by realvictorprm** [comment]
@gusty I agree on this + I'm also aware of this. If you have some time left it would be great to work together on it @gusty 😃

**Comment by gusty** [comment]
@realvictorprm I'm happy to join efforts on this.

I think the first step is to add corner case tests that will help us to see what's working and what's not, so we can work on what's not without breaking anything.

**Comment by jackfoxy** [comment]
Now that 4.5 is out and the span work is done, I hope this makes it to the next release.

**Comment by gerardtoconnor** [comment]
Just to add to edge cases under review:

I was trying to use `PrintfFormat` to type enforce resolution of parsers and it initially appeared to work for `int` but then same approach for `string` did not work... while float did work, so I thought was Value/Ref type issue but then tried `bool` and that didn't work like String.

`int` & `float` work, `string` & `bool` do not!?

(ParseApply methods are dummy implementations for now)

```fsharp
        type System.String  with static member inline ParseApply (path:string) (fn: string -> ^b) : ^b = fn ""
        type System.Int32   with static member inline ParseApply (path:string) (fn: int   -> ^b) : ^b = fn 0
        type System.Double  with static member inline ParseApply (path:string) (fn: float -> ^b) : ^b = fn 0.
        type System.Boolean with static member inline ParseApply (path:string) (fn: bool -> ^b) : ^b = fn true
        
        let inline parser (fmt:PrintfFormat< ^a -> ^b,_,_,^b>) (fn:^a -> ^b) (v:string) : ^b 
            when ^a : (static member ParseApply: string -> (^a -> ^b) -> ^b) =
            (^a : (static member ParseApply: string -> (^a -> ^b) -> ^b)(v,fn))

        let inline patternTest (fmt:PrintfFormat< ^a -> Action< ^T>,_,_,Action< ^T>>) (fn:^a -> Action< ^T>) v : Action< ^T> = parser fmt fn v

        let parseFn1 = patternTest "adfadf%i" (fun v -> printfn "%i" v; Unchecked.defaultof<Action<unit>> ) // works
        let parseFn2 = patternTest "adf%s245" (fun v -> printfn "%s" v; Unchecked.defaultof<Action<unit>> ) // ERROR
        let parseFn3 = patternTest "adfadf%f" (fun v -> printfn "%f" v; Unchecked.defaultof<Action<unit>> ) // works
        let parseFn4 = patternTest "adfadf%b" (fun v -> printfn "%b" v; Unchecked.defaultof<Action<unit>> ) // ERROR
```

The error I get on `result2` function format string input is `The type 'string' does not support the operator 'ParseApply'`, similarly, `result4` error is `The type 'bool' does not support the operator 'ParseApply'`.

Seems to apply to

**Comment by KevinRansom** [comment]
@TobyShaw, @dsyme, this feels like an orphaned PR.  Who will take it over, or can I close it?

**Comment by jackfoxy** [comment]
@KevinRansom  please do not close. I have an interest in seeing this fixed. Unfortunately I don't have the time to come up to speed and attempt fixing it myself right now.

**Comment by KevinRansom** [comment]
@jackfoxy without someone to drive it, it is just repo noise.  I am trying to clean out some of that noise.  I am happy for PRs to remain open while they are being driven to completion.

Kevin

**Comment by jackfoxy** [comment]
@KevinRansom  OK, I can understand that, but I think the language suggestion this references
https://github.com/fsharp/fslang-suggestions/issues/29 is closed for unknown reason. I would like to keep alive "Allow Type Extensions and Extension Methods to Satisfy Constraints" as an approved language feature. I kind of consider it a bug that it does not work.

**Comment by KevinRansom** [comment]
@jack-pappas for sure re-open the language suggestion, or ask for a justification for the closure.

I will keep it open for a bit to see if we can get someone to drive this to completion.  But it will close if no one steps forward.

**Comment by cartermp** [comment]
I re-opened it. It was auto-closed by a tool to migrate from UserVoice, where it was marked as a duplicate of a suggestion that needed the same underlying work.

**Comment by realvictorprm** [comment]
Please don't close this. I really want to take this over as soon as the SRTP bugs are fixed (which unfortunately will take some as it's very difficult).

**Comment by dsyme** [comment]
I'll keep this one open.  It's a significantly important feature and I'll integrate bit by bit

**Comment by KevinRansom** [comment]
@dsyme sweet :-)

**Comment by dsyme** [comment]
Closing in favour of https://github.com/Microsoft/visualfsharp/pull/6286 which has brought this up-to-date re. master

**Comment by dsyme** [comment]
> I was trying to use PrintfFormat to type enforce resolution of parsers and it initially appeared to work for int but then same approach for string did not work... while float did work, so I thought was Value/Ref type issue but then tried bool and that didn't work like String. `int` & `float` work, string & bool do not!?

@gerardtoconnor This is in reply to a really a really old comment, but just wanted to say that I double checked this and none of the four types work.  However type errors are shown in two phases in the IDE, so it's not until you clear the errors for `string` and `bool` that you get shown the errors for `int` and `float`.  This is related to the process of applying defaults.

**Comment by gusty** [comment]
@dsyme I think the process of applying defaults must be reviewed at some point as I'm under the impression that it can be smarter.

Sometimes the compiler applies defaults having a good candidate, I don't have an example right now, but when I was looking at the non-sealed types problem I ended up there. Just my 2 cents.
